blscurve/blst/blst_min_pubkey_sig_core.nim

```diff
@@ -378,3 +381,157 @@ func finish*(ctx: var ContextCoreAggregateVerify, signature: Signature or Aggreg
 
   ctx.commit()
   result = bool ctx.finalVerify()
+
+# Parallelized Batch Verifier primitives
+# ----------------------------------------------------------------------
+#
+# Ultimately this can be merged with the internal ContextCoreAggregateVerify
+# but:
+# - The previous code was audited
+# - For now we only support BLST
+#   though MIRACL is straightforward, context merge is just a FP12 multiplication
+#   since GT is a multiplicative group.
+# - we need to hold cryptographycally secure
+#   random bytes to protect against forged signatures
+#   that would not pass non-aggregated verification
+#   via random scalar blinding
+#   https://ethresear.ch/t/fast-verification-of-multiple-bls-signatures/5407/14
+#
+#
+# Tradeoffs:
+# - Adds:
+#   - a blinding multiplication on G2
+#   - a blinding mul on G1 or Fp
+# - Saves:
+#   - Only 1 Miller Loop per thread (50% pairing)
+#   - Only 1 Final Exponentiation per batch (50% pairing cost)
+#
+# Assuming blinding muls cost 60% of a pairing (worst case with 255-bit blinding)
+# verifying 3 signatures would have a base cost of 300
+# Batched single threaded the cost would be
+# 60*3 (blinding 255-bit) + 50 (Miller) + 50 (final exp) = 280
+#
+# With 64-bit blinding and ~20% overhead
+# (not 15% because no endomorphism acceleration with 64-bit)
+# 20*3 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 160
+#
+# If split on 2 cores, the critical path is
+# 20*2 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 140
+#
+# If split on 3 cores, the critical path is
+# 20*1 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 120
+
+type
+  ContextMultiAggregateVerify*[DomainSepTag: static string] = object
+    ## A context for multi signature verification
+    ##
+    ## This context holds a secure blinding scalar,
+    ## it does not use secret data but it is necessary
+    ## to have data not in the control of an attacker
+    ## to prevent forging valid aggregated signatures
+    ## from 2 invalid individual signatures using
+    ## the bilinearity property of pairings.
+    ## https://ethresear.ch/t/fast-verification-of-multiple-bls-signatures/5407/14
+    c: blst_pairing
+    secureBlinding: array[32, byte]
+
+func init*[T: char|byte](
+       ctx: var ContextMultiAggregateVerify,
+       secureRandomBytes: array[32, byte],
+       threadSepTag: openarray[T]
+     ) {.inline.} =
```

Comment:
why inline?

Reply:
When it's a single function call or just a copy, inline will give the compiler visibility to optimize away things like

`ctx.secureBlinding = secureRandomBytes`

That said, due to the threadSepTag issue above this will become

```Nim
# pseudo code
var shaCtx: SHA256
shaCtx.init()
shaCtx.update(secureRandomBytes)
shaCtx.update(threadSepTag)
shaCtx.finish(ctx.secureBlinding)
```
and so won't be inline anymore.